source/common/network/connection_impl.cc

```diff
@@ -270,6 +270,13 @@ void ConnectionImpl::readDisable(bool disable) {
 }
 
 void ConnectionImpl::raiseEvent(ConnectionEvent event) {
+  // We may have pending data in the write buffer on transport handshake
+  // completion, which may also have completed in the context of onReadReady(),
+  // where no check of the write buffer is made. Provide an opportunity to flush
+  // here. If connection write is not ready, this is harmless.
+  if (event == ConnectionEvent::Connected && write_buffer_->length() > 0) {
+    onWriteReady();
```

Comment:
onWriteReady() can close the socket.  Need to check that before continuing.  Although, if that happens, filters will get notified of a close before being notified that it was connected.

Hmmm... Thinking this through, it is probably better to notify all filters that they're connected (ie put it back to the way it was before), and make this case be:

```
if (state() == State::Open && event == ConnectionEvent::Connected && write_buffer_->length() > 0)
```

Reply:
Sorry to go in circles on this; ConnectionImpl is complicated.

Reply:
Sure, this is an area of the code I'm less familiar with, and it's hard to reason about what callbacks can and can't do here, so thanks for the patience.